ugc/storage.cpp

```diff
@@ -1,33 +1,434 @@
+  string data;
+  auto const indexFilePath = GetIndexFilePath();
+  try
+  {
+    FileReader r(indexFilePath);
+    r.ReadAsString(data);
+  }
+  catch (RootException const & exception)
+  {
+    LOG(LWARNING, ("Exception while reading file:", indexFilePath));
+    return;
+  }
+
+  DeserializeUGCIndex(data, m_UGCIndexes, m_numberOfDeleted);
+  sort(m_UGCIndexes.begin(), m_UGCIndexes.end(), [](UGCIndex const & lhs, UGCIndex const & rhs) -> bool {
+    return lhs.m_offset < rhs.m_offset;
+  });
+}
+
+void Storage::SaveIndex() const
+{
+  ASSERT_THREAD_CHECKER(m_threadChecker, ());
+  auto const jsonData = SerializeUGCIndex(m_UGCIndexes);
+  auto const indexFilePath = GetIndexFilePath();
+  try
+  {
+    FileWriter w(indexFilePath);
+    w.Write(jsonData.c_str(), jsonData.length());
+  }
+  catch (RootException const & exception)
+  {
+    LOG(LWARNING, ("Exception while writing file:", indexFilePath));
+  }
+}
+
+void Storage::Defragmentation()
+{
+  ASSERT_THREAD_CHECKER(m_threadChecker, ());
+  auto const indexesSize = m_UGCIndexes.size();
+  if (m_numberOfDeleted < indexesSize / 2)
+    return;
+
+  auto const ugcFilePath = GetUGCFilePath();
+  auto const tmpUGCFilePath = ugcFilePath + kTmpFileExtension;
+  FileReader r(ugcFilePath);
+  vector<uint8_t> buf;
+  for (size_t i = 0; i < indexesSize; ++i)
+  {
+    auto const & index = m_UGCIndexes[i];
+    if (index.m_isDeleted)
+      continue;
+
+    auto const offset = index.m_offset;
+    uint64_t nextOffset;
+    if (i == indexesSize - 1)
+      GetUGCFileSize(nextOffset);
+    else
+      nextOffset = m_UGCIndexes[i + 1].m_offset;
+
+    auto const bufSize = nextOffset - offset;
+    try
+    {
+      r.Read(offset, buf.data(), bufSize);
+    }
+    catch (RootException const & excpetion)
+    {
+      LOG(LWARNING, ("Exception while reading file:", ugcFilePath));
+      return;
+    }
+  }
+
+  try
+  {
+    FileWriter w(tmpUGCFilePath);
+    w.Write(buf.data(), buf.size());
+  }
+  catch (RootException const & exception)
+  {
+    LOG(LWARNING, ("Exception while reading file:", tmpUGCFilePath));
+    return;
+  }
+
+  m_UGCIndexes.erase(remove_if(m_UGCIndexes.begin(), m_UGCIndexes.end(), [](UGCIndex const & i) -> bool {
+    return i.m_isDeleted;
+  }));
+
+  CHECK(my::DeleteFileX(ugcFilePath), ());
+  CHECK(my::RenameFileX(tmpUGCFilePath, ugcFilePath), ());
+
+  m_numberOfDeleted = 0;
+}
+
+string Storage::GetUGCToSend() const
+{
+  ASSERT_THREAD_CHECKER(m_threadChecker, ());
+  if (m_UGCIndexes.empty())
+    return string();
+
+  auto array = my::NewJSONArray();
+  auto const indexesSize = m_UGCIndexes.size();
+  auto const ugcFilePath = GetUGCFilePath();
+  FileReader r(ugcFilePath);
+  vector<uint8_t> buf;
+  for (size_t i = 0; i < indexesSize; ++i)
+  {
+    buf.clear();
+    auto const & index = m_UGCIndexes[i];
+    if (index.m_isSynchronized)
+      continue;
+
+    auto const offset = index.m_offset;
+    uint64_t nextOffset;
+    if (i == indexesSize - 1)
```

Comment:
I saw this code several times, could you please implement a shortcut for de-duplication?